src/Voron/Impl/LowLevelTransaction.cs

```diff
@@ -1440,7 +1445,66 @@ internal void OnAfterCommitWhenNewTransactionsPrevented()
             AfterCommitWhenNewTransactionsPrevented?.Invoke(this);
         }
 
+#if DEBUG
+        //overflowPageId, Parent
+        private readonly Dictionary<long, long> _overflowPagesToBeRemoved = new();
+        [Conditional("DEBUG")]
+        private void ValidateOverflowPagesRemoval()
+        {
+            if (_overflowPagesToBeRemoved.Count == 0)
+                return;
+            
+            StringBuilder sb = new();
+            var pagesMissed = from overflowPage in _overflowPagesToBeRemoved
+                group overflowPage.Key by overflowPage.Value
+                into g
+                select new {
+                    Parent = g.Key, 
+                    OverflowPages = g.Select(i => i).ToArray()
+                };
+
+            sb.AppendLine("We commit but we still have not deleted all overflow pages to retrieve used storage. Missed pages: ");
+            foreach (var pageMissed in pagesMissed)
+            {
+                sb.AppendLine($"Parent: {pageMissed.Parent}, Overflow pages: {string.Join(",", pageMissed.OverflowPages)}");
+            }
+
+            VoronUnrecoverableErrorException.Raise(_env, sb.ToString());
+        }
+
+        [Conditional("DEBUG")]
+        private void TrackOverflowPageRemoval(long pageId)
+        {
+            if (_overflowPagesToBeRemoved.ContainsKey(pageId))
+            {
+                _overflowPagesToBeRemoved.Remove(pageId);
+                return;
+            }
+            
+            if (_state.Root.PageCount < pageId)
+                return;
```

Comment:
@ayende 
When the page is not allocated `FastTests.Voron.Bugs.RavenDB_16464.ShouldFreeAllocationsInScratchBuffersSoTheyCanBeCleaned`) we've to ensure it's already allocated to avoid exception. Is this valid?

Reply:
Yes, this is there to protect us from missing this the next time. Doesn't have to be 100% accurate.
And for this, we usually allocate the first few pages for metadata, etc.